crates/zkevm_circuits/src/bn254/ec_mul/implementation.rs

```diff
@@ -0,0 +1,428 @@
+    let zero_var = Num::allocated_constant(cs, F::ZERO).get_variable();
+    let mut limbs = [zero_var; N];
+    assert!(N >= 16);
+
+    for (dst, src) in limbs.array_chunks_mut::<2>().zip(value.inner.iter()) {
+        let [b0, b1, b2, b3] = src.to_le_bytes(cs);
+        let low = UInt16::from_le_bytes(cs, [b0, b1]);
+        let high = UInt16::from_le_bytes(cs, [b2, b3]);
+
+        *dst = [low.get_variable(), high.get_variable()];
+    }
+
+    let mut max_value = U1024::from_word(1u64);
+    max_value = max_value.shl_vartime(256);
+    max_value = max_value.saturating_sub(&U1024::from_word(1u64));
+
+    let (overflows, rem) = max_value.div_rem(&params.modulus_u1024);
+    let mut max_moduluses = overflows.as_words()[0] as u32;
+    if rem.is_zero().unwrap_u8() != 1 {
+        max_moduluses += 1;
+    }
+
+    let element = NonNativeFieldOverU16 {
+        limbs,
+        non_zero_limbs: 16,
+        tracker: OverflowTracker { max_moduluses },
+        form: RepresentationForm::Normalized,
+        params: params.clone(),
+        _marker: std::marker::PhantomData,
+    };
+
+    element
+}
+
+/// Converts the non-native field eelement over `u16` to a `UInt256`.
+/// Note that caller must ensure that the field element is normalized,
+/// otherwise this will fail.
+pub fn convert_field_element_to_uint256<F, CS, P, const N: usize>(
+    cs: &mut CS,
+    mut value: NonNativeFieldOverU16<F, P, N>,
+) -> UInt256<F>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+    P: PrimeField,
+{
+    assert_eq!(value.form, RepresentationForm::Normalized);
+    assert_eq!(value.tracker.max_moduluses, 1);
+
+    let mut limbs = [UInt32::<F>::zero(cs); 8];
+    let two_pow_16 = Num::allocated_constant(cs, F::from_u64_unchecked(2u32.pow(16) as u64));
+    for (dst, src) in limbs.iter_mut().zip(value.limbs.array_chunks_mut::<2>()) {
+        let low = Num::from_variable(src[0]);
+        let high = Num::from_variable(src[1]);
+        *dst = unsafe {
+            UInt32::from_variable_unchecked(
+                Num::fma(cs, &high, &two_pow_16, &F::ONE, &low, &F::ONE).get_variable(),
+            )
+        };
+    }
+
+    UInt256 { inner: limbs }
+}
+
+#[derive(Debug, Clone)]
+pub struct ScalarDecomposition<F: SmallField> {
+    pub k1: BN256ScalarNNField<F>,
+    pub k2: BN256ScalarNNField<F>,
+    pub k1_was_negated: Boolean<F>,
+    pub k2_was_negated: Boolean<F>,
+}
+
+impl<F> ScalarDecomposition<F>
+where
+    F: SmallField,
+{
+    fn u256_from_hex_str<CS>(cs: &mut CS, s: &str) -> UInt256<F>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let v = U256::from_str_radix(s, 16).unwrap();
+        UInt256::allocated_constant(cs, v)
+    }
+
+    fn bigint_from_hex_str<CS>(cs: &mut CS, s: &str) -> UInt512<F>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let v = U256::from_str_radix(s, 16).unwrap();
+        UInt512::allocated_constant(cs, (v, U256::zero()))
+    }
+
+    pub fn from<CS>(
```

Comment:
It could be a huge optimization here. Rewrite this without cs. put k1, k2 like a witness and check  GLV multiplication k = k_1 +k_2 * lambda mod n